prog/minio/minio_server_nexus.rb

```diff
@@ -95,9 +102,23 @@ def cluster
       decr_restart
       push self.class, frame, "minio_restart"
     end
+
+    when_refresh_certificates_set? do
+      hop_refresh_certificates
+    end
+
     nap 10
   end
 
+  label def refresh_certificates
+    decr_refresh_certificates
+    cert, cert_key = create_certificate
```

Comment:
So each time we refresh the root certificate, we also refresh the server certificate, irrelevant from its expiration date.

Recreating the server certificate when it is close to be expired (can be done in cluster's refresh cert) and just putting certs to the VM (here or in reconfigure) should be enough, no? Alternatively, if you want to refresh the server certificates here, then you can have a similar check here and don't check the expiration in cluster's refresh cert. Though, that would be even more complicated.

I suspect creating the cert in cluster's refresh and then just increasing server's reconfigure would be enough.

Reply:
I believe servers need to create their own certificates. The main reason being, new servers may need to be added to the cluster and then it would require a whole new complex ping-ponging with the cluster. I have an alternative solution, I'm thinking that instead of refreshing the certificates after cluster cert rotation, we can simply increment another semaphore that would rewrite the ca bundle and servers independently recreate certs close to expiration.

Reply:
I suppose you need two passes, right? One to ensure that new trust roots are distributed, then to change the cert presented by a node.

Reply:
That's now fixed. If the trust roots are refreshed, we reconfigure, which puts the new ca_bundle into the servers. If it's time to refresh the individual server cert, we refresh and reconfigure again.